### ISSUE TITLE
Notes about architecture and structure

### ISSUE THREAD

**Issue by Birkbjo** [dhis2/app-hub#139]
I've been working with the backend the last few weeks, and in that regard I have some notes and feedback I want to discuss.

- Model concept
Currently we have no concept of database "models", they are merely validations. I don't know if we need these, but I've been fiddling with the idea of having and "internal" model, which is basically a combination of a Joi-schema and some helper methods and constants for that "model". Does not need to be a class, just a file exporting helpers for that entity. Like having a map of internal fields to database fields, and a way to parse from and to the database. These could for instance be in root of 'models/v2'.

- The 'data' concept
   I feel these cumbersome to use, as we have to parse the data anytime it's used. I feel like these can be grouped as a 'service' for each entity, and in combination with the 'models', it would be easier to use internally. With this I mean that we would never 'expose' the result of the database queries directly, they would be parsed to an internal structure, so we don't need to care about column names and collecting joined relations anytime data is fetched.

- Validation
 Currently we are validating in both routes and 'data', I belive this is overkill. It should probably be up to the route handlers to do validation through hapi. Internal wrong usage is developer error and we shouldn't use Joi for type-checking.

- Transactions
  I've been fiddling with the 'service'-idea, and I believe if we do this right we can cut down on the 'try-catching' of transactions. Knex supports using transactions in the same way you would've used `knex()`, so we can use these interchangeably. This means that we can use the transaction-handler syntax in route-handlers and compose service-methods by passing either the transaction or `knex`-object - with the benefit of automatic rollback. Rough concept:
```javascript 
// some handler
const createVersionAndAddToChannel = (trx => {
     const v = await Version.create(payload.version, trx);
     await Channel.addVersion({appVersionId: v.id, channelName: payload.channel, ...}, trx)
     return v;
})
const version = h.db.transaction(createVersionAndAddToChannel);
```

We probably still need some explicit transaction management due to the normalized structure of our database, and some entities never 'live alone'.

- Error handling
 I believe this can be simplified as well. Services or data-fetching does not necessarily need to 'try-catch' and check for everything. It's actually a good thing that they don't handle the errors if we use the transaction-handler idea. We could also probably use something like https://github.com/Vincit/db-errors with `onPreResponse`-hapi hook to format and check for different errors, and return a meaningful error message to the client. 

- Uuid and id seperation
This is causing some headache and potential performance loss. In simple cases where you just need ids to add a relation or insert something, you now need a query for each entity as we need the id to perform inserts. Ie. if I want to add a user to an organisation I would just need the userId and organsiationId. However, since our join tables are based on ids and we just have the uuids on the client, I would need to run a query to grab the organisation and get the id, and the same thing with the user (well not really - we do have this from the auth in the request). Feels wrong to me.

**Comment by erikarenhill** [comment]
Thanks for the thorough analysis and input @Birkbjo 
Here's my thoughts on the subject

## Model + data concept:
Reading in between the lines this to me sounds like you would be in favor for an ORM solution?

I generally think that using an ORM lead to more "magic" code that can be hard to follow, or end up being too generic leading up to making small changes hard or time-consuming. They also tend to produce complex bad performant SQL-queries for the most simple things.

## Validation:
Route validation can be performed both on incoming and outgoing data, making sure that a code change doesn't break the format of a response for example. 

I'm generally in favor of catching errors in an early stage if possible, at the same time every "layer" of the architecture should if possible prevent any errors in between. 

Usually I consider each layer of an application as an API towards eachother where any input coming from a layer on top should be validated when pushed downwards (for example from a route handler to the data layer). It's more work but consider the case that one developer takes responsibility for the data layer and expose features for another developer working on the business logic and a third one on the API and the route handlers there should be some validation and error handling in between those, this also makes isolated testing easier and the validation acts as a contract between different layers.

We have a previous decision about validation of config objects (as used in the data layer today), and I think Joi is an excellent library for resolving this: https://github.com/dhis2/notes/pull/41/files

It removes lots of manual logic when we can add required, allow null, specific string lengths, allowed values in arrays etc..


## UUID/ID separation:
I don't think we'll ever see this as a real performance issue as the app store API as it won't contain millions of rows or will be a very write-intense service.

I agree that we could move the UUID to be the primary key, that would potentially make it easier to merge different databases without having a key-collision. It would also make clustering and synchronizing databases easier.

We initially had a discussion with a decision about not exposing the primary keys from the database that lead to the design we have today. Probably it would be doable to redesign and remove the incremental ids in a day or so. 

Revisiting this today I can't really find a proper argument for keeping the incremental ID's. Maybe we should however revisit the question regarding if we want to expose internal database id's and re-use them as external.

**Comment by erikarenhill** [comment]
We should also consider flattening project structure on the file system to avoid having so deep relative paths such as 
`const createLocalizedAppVersion = require('../../../../data/createLocalizedAppVersion')`

@varl had some idea a while ago about this, please fill in.

**Comment by Birkbjo** [comment]
> Reading in between the lines this to me sounds like you would be in favor for an ORM solution?

Not necessarily, and I'm not proposing adding an ORM-library. It would be very light, and would basically just mean that we have a map between an internal structure and the database-columns - so the data can be mapped whenever something is grabbed from the database. I find it very annoying to handle the normalized structure of the data whenever a 'data'-fetch is done. Transformation of the data should in my opinion be done by the same function that queries the database, however we could use the existing simple 'data'-functions but have a service-layer above that transform the data to an internal structure. You would still do your own queries.

> We have a previous decision about validation of config objects (as used in the data layer today), and I think Joi is an excellent library for resolving this: https://github.com/dhis2/notes/pull/41/files

I would think this was just for exposed API-functions. If we would do this everywhere we would basically use Joi as a type-checker, and then we could rather look into typescript.
Your counterpoint of the layers basically being exposed APIs is valid though, however right now the 'layers' are tightly coupled, as the route-handlers deal with the raw database output. 
This is my main concern along with the huge handler-functions and something I wanted to solve with 'services' that transform the data into an internal structure so it's easier for business logic and route handlers to use the data. This would also mean that it would be easier to change database schema down the line.


> I don't think we'll ever see this as a real performance issue as the app store API as it won't contain millions of rows or will be a very write-intense service.

Right, I'm just stuck in the performance mindset all the time, and I guess that's the origin of my objection to multiple runtime validations of the same object.

I will open a draft-PR to show my experiments. Note that the PR will have some other functionality I have explored (like filtering and paging).

**Comment by erikarenhill** [comment]
> > Reading in between the lines this to me sounds like you would be in favor for an ORM solution?
> 
> Not necessarily, and I'm not proposing adding an ORM-library. It would be very light, and would basically just mean that we have a map between an internal structure and the database-columns - so the data can be mapped whenever something is grabbed from the database. I find it very annoying to handle the normalized structure of the data whenever a 'data'-fetch is done. Transformation of the data should in my opinion be done by the same function that queries the database, however we could use the existing simple 'data'-functions but have a service-layer above that transform the data to an internal structure. You would still do your own queries.
> 

I agree that it would be nice to avoid having to work with the database result-sets. 

> > We have a previous decision about validation of config objects (as used in the data layer today), and I think Joi is an excellent library for resolving this: https://github.com/dhis2/notes/pull/41/files
> 
> I would think this was just for exposed API-functions. If we would do this everywhere we would basically use Joi as a type-checker, and then we should rather look into typescript.

You got a point but I don't entirely agree since joi can also check for string lengths (ok TS can kind of solve this too by using phantom types), validate url, e-mail, allow specific values in arrays etc.. and it does this in runtime and not in compile time. But let's bring this up for a discussion in a meeting again then?

> Your counterpoint of the layers basically being exposed APIs is valid though, however right now the 'layers' are tightly coupled, as the route-handlers deal with the raw database output.
I think this has to change, moving forward we should iterate the architechture and move that out of the handlers.

> This is my main concern along with the huge handler-functions and something I wanted to solve with 'services' that transform the data into an internal structure so it's easier for business logic and route handlers to use the data.
> 
> > I don't think we'll ever see this as a real performance issue as the app store API as it won't contain millions of rows or will be a very write-intense service.
> 
> Right, I'm just stuck in the performance mindset all the time, and I guess that's the origin of my objection to multiple runtime validations of the same object.

I believe it's great to always consider performance, but to a certain degree, otherwise we could have denormalized all tables also in the database if we just wanted the best performance possible to fetch data without any joins

> 
> I will open a draft-PR to show my experiments. Note that the PR will have some other functionality I have explored (like filtering and paging).

Looking forward to see that!

**Comment by erikarenhill** [comment]
@varl me and @Birkbjo is considering dropping the numeric id in favor of just using uuid for primary keys, any objections against that? 

We'll then use the same id both internally and externally which makes the code easier, less db queries looking up id vs uuid etc. Regarding security by obscurity with not exposing internal id's which we discussed in the project startup I believe using uuid v4 is hard/impossible enough to guess a series of with rows for

**Comment by varl** [comment]
I agree with dropping **id** in favor of **uuid**. As long as we use UUID v4 then it is exceedingly hard to guess.

We need migration scripts for both up and down as usual.
